### ISSUE TITLE
Unexpected Application Error when deployed on Kubernetes behind an ingress reverse proxy

### ISSUE THREAD

**Issue by mvysny** [vaadin/hilla#899]
### Description of the bug 

I'm running the basic Hilla 2.0+React example app generated via [Hilla quick start](https://hilla.dev/docs/react/start/quick). Deploying on Microk8s Kubernetes, behind an Ingress reverse proxy with path rewriting. The Hilla app doesn't load and instead fails with:
```
Unexpected Application Error!
404 Not Found
```
![image](https://user-images.githubusercontent.com/254091/232697204-1496671d-8167-4bfa-b253-e8cef3171a70.png)

### Minimal reproducible example

Please visit https://v-herd.eu/my-hilla-app/ to immediately observe the behavior. I'll attach the Kubernetes yaml below.

The Dockerfile with all instructions on how to build the image is at https://github.com/mvysny/my-hilla-app . The image works correctly when running via `docker run --rm -ti -p8080:8080 test/my-hilla-app`; the failure can be observed only when running behind the Ingress reverse proxy.

### Expected behavior

The Hilla app should work properly and should render the welcome screen.

### Actual behavior

Hilla app fails in the browser. On server everything looks okay:
```
                        _     _ _ _                               
  _ __ ___  _   _      | |__ (_) | | __ _        __ _ _ __  _ __  
 | '_ ` _ \| | | |_____| '_ \| | | |/ _` |_____ / _` | '_ \| '_ \ 
 | | | | | | |_| |_____| | | | | | | (_| |_____| (_| | |_) | |_) |
 |_| |_| |_|\__, |     |_| |_|_|_|_|\__,_|      \__,_| .__/| .__/ 
            |___/                                    |_|   |_|    
2023-04-18T06:56:13.558Z  INFO 7 --- [           main] com.example.application.Application      : Starting Application v1.0-SNAPSHOT using Java 17.0.6 with PID 7 (/app/my-hilla-app-1.0-SNAPSHOT.jar started by root in /app)
2023-04-18T06:56:13.622Z  INFO 7 --- [           main] com.example.application.Application      : No active profile set, falling back to 1 default profile: "default"
2023-04-18T06:56:19.354Z  INFO 7 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
2023-04-18T06:56:19.441Z  INFO 7 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2023-04-18T06:56:19.442Z  INFO 7 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.7]
2023-04-18T06:56:20.045Z  INFO 7 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2023-04-18T06:56:20.048Z  INFO 7 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 6226 ms
2023-04-18T06:56:22.136Z  INFO 7 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing AtmosphereFramework
2023-04-18T06:56:22.451Z  INFO 7 --- [           main] c.v.f.s.VaadinServletContextInitializer  : Search for VaadinAppShell took 98 ms
2023-04-18T06:56:24.423Z  INFO 7 --- [           main] dev.hilla.EndpointRegistry               : Registered endpoint 'HelloReactEndpoint' with class 'class com.example.application.endpoints.helloreact.HelloReactEndpoint'
2023-04-18T06:56:25.225Z  INFO 7 --- [           main] c.v.f.s.DefaultDeploymentConfiguration   : Vaadin is running in production mode.
2023-04-18T06:56:25.432Z  INFO 7 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2023-04-18T06:56:25.457Z  INFO 7 --- [           main] com.example.application.Application      : Started Application in 13.923 seconds (process running for 16.083)
2023-04-18T06:57:54.756Z  INFO 7 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2023-04-18T06:57:54.757Z  INFO 7 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2023-04-18T06:57:54.759Z  INFO 7 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 2 ms
2023-04-18T06:57:54.944Z  INFO 7 --- [nio-8080-exec-1] c.vaadin.flow.spring.SpringInstantiator  : The number of beans implementing 'I18NProvider' is 0. Cannot use Spring beans for I18N, falling back to the default behavior
```

The Browser JavaScript console is also quite empty, no interesting errors there:

```
Tue Apr 18 2023 10:08:39 GMT+0300 (Eastern European Summer Time) Atmosphere: unload event [indexhtml-200958ac.js:2280:38943]
GET https://v-herd.eu/favicon.ico [HTTP/2 404 Not Found 10ms]
```

### Versions:
    - Vaadin / Hilla version: 2.0.4
    - Node version: v18.14.1
    - Java version: 17-temurin
    - OS version: Linux in Docker
    - Browser version (if applicable): Firefox 112.0.1
    - Application Server (if applicable): Spring Boot

**Comment by mvysny** [comment]
kubernetes yaml config file:
```yaml
#                                                                                                                                                                                             
# Microk8s resource config file for my-hilla-app                                                                                                                                              
#                                                                                                                                                                                             
                                                                                                                                                                                              
apiVersion: v1                                                                                 
kind: Namespace                                                                                
metadata:                                      
  name: shepherd-my-hilla-app                  
---                                            
apiVersion: apps/v1                            
kind: Deployment                               
metadata:                                      
  name: deployment                             
  namespace: shepherd-my-hilla-app                                                             
spec:                                                                                          
  selector:                                    
    matchLabels:                               
      app: pod                                 
  template:                                    
    metadata:                                  
      labels:                                  
        app: pod                                                                               
    spec:                                      
      containers:
      - name: main
        image: localhost:32000/shepherd/my-hilla-app@sha256:d2f5c33de109ddad996fc5f8a0535463d92010271e039e689e95484b714f1832
        ports:
        - containerPort: 8080
        resources:
          requests:
            memory: "64Mi"
            cpu: 0
          limits:
            memory: "256Mi"  # https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
            cpu: 1
---
apiVersion: v1
kind: Service
metadata:
  name: service
  namespace: shepherd-my-hilla-app
spec:
  selector:
    app: pod
  ports:
    - port: 8080
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ingress-main
  namespace: shepherd-my-hilla-app
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /$3
    nginx.ingress.kubernetes.io/proxy-cookie-path: / /$1
    nginx.ingress.kubernetes.io/configuration-snippet: |
      rewrite ^(/my-hilla-app)$ $1/ permanent;
spec:
  tls:
  - hosts:
    - v-herd.eu
  rules:
    - host: v-herd.eu
      http:
        paths:
          - path: /(my-hilla-app)(/|$)(.*)
            pathType: Prefix
            backend:
              service:
                name: service
                port:
                  number: 8080
```

**Comment by mvysny** [comment]
Thank you, I've added the line to `frontend/index.html`. Hilla now downloads the favicon from a different location: `https://v-herd.eu/my-hilla-app/`, unfortunately it's still a 404 and Hilla fails with the same error: Unexpected Application Error! 404 Not Found

![image](https://user-images.githubusercontent.com/254091/232702517-ada2d795-d86e-4aac-b338-10a7fe9e3cdb.png)

**Comment by cromoteca** [comment]
Now it's trying to download JSON instead. Could you try adding an icon to see if it can be found correctly?

**Comment by mvysny** [comment]
I've added the favicon.ico and the Network tab now shows 200 OK for the favicon, unfortunately the Hilla app still fails with the same error: Unexpected Application Error! 404 Not Found

![image](https://user-images.githubusercontent.com/254091/232708576-d237f638-de1a-475c-ad04-acd6844a2c85.png)

**Comment by cromoteca** [comment]
So that's not related to favicon. The error message comes from the React router. Some [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel) is not working. To be honest, I can't think of any reason for that, as this should happen on the client side. The server fulfilled all the requests.

**Comment by mvysny** [comment]
The solution is to add `basename: new URL(document.baseURI).pathname` to your `frontend/routes.tsx` as per the [React Router basename documentation](https://reactrouter.com/en/main/routers/create-browser-router#basename):

```
...
const router = createBrowserRouter([...routes], {basename: new URL(document.baseURI).pathname });
...
```

**Comment by Artur-** [comment]
Should this be added to the default starter?

**Comment by Legioth** [comment]
...and now in the age of an automatically generated routing configuration, it should probably be added to that generated file as well.

**Comment by Legioth** [comment]
...never mind. Seems like it's already there: https://github.com/vaadin/hilla/blob/512117b807ef523efbb1a25eb4dec4972ed20119/packages/ts/file-router/src/runtime/RouterConfigurationBuilder.ts#L229

**Comment by Legioth** [comment]
Fixed in Vaadin 24.4 by https://github.com/vaadin/hilla/commit/6e3d0fb7386d4357ceb22dabc83d3a7e84bfc601 when using the router config builder.
